File: algorithms/rs/242.有效的字母异位词.rs
/*
 * @lc app=leetcode.cn id=242 lang=rust
 *
 * [242] 有效的字母异位词
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        let mut map = vec![0; 26];
        for ch in s.chars() {
            map[ch as usize - 97] += 1;
        }
        for ch in t.chars() {
            map[ch as usize - 97] -= 1;
        }
        for n in map {
            if (n != 0) {
                return false;
            }
        }
        true
    }
}
// @lc code=end

File: algorithms/golang/242.有效的字母异位词.go
/*
 * @lc app=leetcode.cn id=242 lang=golang
 *
 * [242] 有效的字母异位词
 */

/**
 * 排序后比较
 * 时间复杂度 n*O(n)
 * 空间复杂度 O(log(n)) - 排序空间复杂度为 O(log(n))
 */
func isAnagram_1(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	arr := []byte(s)
	sort.Slice(arr, func(i, j int) bool { return arr[i] > arr[j] })
	s = string(arr)
	arr = []byte(t)
	sort.Slice(arr, func(i, j int) bool { return arr[i] > arr[j] })
	t = string(arr)
	return s == t
}

// @lc code=start
/**
 * 数组记录字符出现次数
 * 时间复杂度 O(n)
 * 空间复杂度 O(26)
 */
func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	arr := [26]int{}
	for i := 0; i < len(s); i++ {
		arr[byte(s[i])-'a']++
		arr[byte(t[i])-'a']--
	}
	for _, n := range arr {
		if n != 0 {
			return false
		}
	}
	return true
}

// @lc code=end